elasticsearch/security_api.go

```diff
@@ -0,0 +1,212 @@
+		Application: string(privilegesReq.Application),
+		Name:        string(privilegesReq.Privilege),
+	}
+	err := doRequest(client, req, &confirmation)
+	return confirmation, err
+}
+
+func HasPrivileges(client Client, privileges HasPrivilegesRequest, credentials string) (HasPrivilegesResponse, error) {
+	var info HasPrivilegesResponse
+	req := esapi.SecurityHasPrivilegesRequest{Body: esutil.NewJSONReader(privileges)}
+	if credentials != "" {
+		header := make(http.Header)
+		header.Set("Authorization", "ApiKey "+credentials)
+		req.Header = header
+	}
+	err := doRequest(client, req, &info)
+	return info, err
+}
+
+type CreateAPIKeyRequest struct {
+	Name            string         `json:"name"`
+	Expiration      *string        `json:"expiration,omitempty"`
+	RoleDescriptors RoleDescriptor `json:"role_descriptors"`
+}
+
+type CreateAPIKeyResponse struct {
+	APIKey
+	Key string `json:"api_key"`
+}
+
+type GetAPIKeyRequest struct {
+	APIKeyQuery
+	Owner bool `json:"owner"`
+}
+
+type GetAPIKeyResponse struct {
+	APIKeys []APIKeyResponse `json:"api_keys"`
+}
+
+type CreatePrivilegesRequest map[AppName]PrivilegeGroup
+
+type CreatePrivilegesResponse map[AppName]PrivilegeResponse
+
+type HasPrivilegesRequest struct {
+	// can't reuse the `Applications` type because here the JSON attribute must be singular
+	Applications []Application `json:"application"`
+}
+type HasPrivilegesResponse struct {
+	Username    string                             `json:"username"`
+	HasAll      bool                               `json:"has_all_requested"`
+	Application map[AppName]PermissionsPerResource `json:"application"`
+}
+
+type InvalidateAPIKeyRequest struct {
+	APIKeyQuery
+}
+
+type InvalidateAPIKeyResponse struct {
+	Invalidated []string `json:"invalidated_api_keys"`
+	ErrorCount  int      `json:"error_count"`
+}
+
+type DeletePrivilegeRequest struct {
+	Application AppName       `json:"application"`
+	Privilege   PrivilegeName `json:"privilege"`
+}
+
+type DeletePrivilegeResponse map[AppName](map[PrivilegeName]DeleteResponse)
+
+type RoleDescriptor map[AppName]Applications
+
+type Applications struct {
+	Applications []Application `json:"applications"`
+}
+
+type Application struct {
+	Name       AppName           `json:"application"`
+	Privileges []PrivilegeAction `json:"privileges"`
+	Resources  []Resource        `json:"resources"`
+}
+
+type APIKeyResponse struct {
+	APIKey
+	Creation    int64  `json:"creation"`
+	Invalidated bool   `json:"invalidated"`
+	Username    string `json:"username"`
+}
+
+type APIKeyQuery struct {
+	// normally the Elasticsearch API will require either Id or Name, but not both
+	ID   *string `json:"id,omitempty"`
+	Name *string `json:"name,omitempty"`
+}
+
+type APIKey struct {
+	ID           string `json:"id"`
+	Name         string `json:"name"`
+	ExpirationMs *int64 `json:"expiration,omitempty"`
+	// This attribute does not come from Elasticsearch, but is filled in by APM Server
+	Credentials *string `json:"credentials,omitempty"`
```

Comment:
Since there is only 1 usage, I thought adding the field directly to the type would be simpler, and it could be moved out later if it turns out to be problematic?

Reply:
I would classify that as easy, not simple. It's already been problematic in that I had to check what's using it and what's not, and why.

I'm not sure that we should be directly encoding these types for the CLI anyway, but that's a debate for another time.